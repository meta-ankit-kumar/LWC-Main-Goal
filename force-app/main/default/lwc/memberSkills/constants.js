export const MESSAGE = {
	MEMBER_CREATION: {
		SUCCESSFULL: 'New Memeber: MEMBER_NAME is added successfully',
		ERROR: {
			GENERIC: 'There was some problem while retrieving the data: ERRORS'
		}
	}
}