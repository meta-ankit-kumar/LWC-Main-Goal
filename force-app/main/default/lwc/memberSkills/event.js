export class MemberAddedEvent extends CustomEvent {
	static type = 'memberadded';
	constructor(value) {
		super(MemberAddedEvent.type, {
			detail: value
		})
	}
}