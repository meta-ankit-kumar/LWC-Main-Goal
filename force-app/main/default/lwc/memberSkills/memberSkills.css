.name-info {
	display: flex;
	justify-content: center;
}

.member-skills-container {
	display: flex;
    justify-content: center;
    padding: 10px;
}

input, select {
	margin-left: 1.5em;
	display: flex;
	width: 100%;
}

label {
	display: block;
}

.member-skills-container div label, input, select {
    margin-top: 10px;
}

.submit-button {
	width: 60%;
    margin-left: 0px;
    text-align: center;
    display: flex;
    align-items: center;
    justify-content: center;
    font-size: 1.1rem;
}

.message-container {
	overflow: hidden;
    width: 45%;
    margin: 0 auto;
    background: white;
    padding: 10px;
}

.main-member-container {
	box-shadow: 0px 0px 1px 0px;
    border-radius: 1%;
    background: white;
	width: 60%;
    margin: 0 auto;
}