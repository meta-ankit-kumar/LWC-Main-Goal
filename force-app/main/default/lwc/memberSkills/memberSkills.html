<template>
    <div class="main-member-container">
        <div class="member-skills-container">
            <div if:false={errors} class="labels-container">
                <label for="member-name">Name:</label>
                <label for="member-name">Team:</label>
                <label for="member-skills">Skills:</label>
            </div>
            <div if:false={errors} class="inputs-container">
                <form onsubmit={handleSubmit}>
                    <input required minlength="4" type="text" name="member-name" id="member-name">
                    <select required name="member-team-name" onchange={handleTeamNameChange}>
                        <option for:each={teams} for:item="team" key={team.id} value={team.id}>{team.name}</option>
                    </select>
                    <input required type="text" name="member-skills" id="member-skills">
                    <input class="submit-button" type="submit" value="Submit">
                </form>
            </div>
            <div if:true={errors}>
                <p>{errorsToBeShown}</p>
            </div>
        </div>
        <div class="message-container" if:true={message}>
            <c-message text={message} onclearmessage={handleClearMessage}></c-message>
        </div>
    </div>
</template>