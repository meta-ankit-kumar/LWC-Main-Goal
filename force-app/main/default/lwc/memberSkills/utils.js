export const useTemplate = (templateString, params) => {
	Object.entries(params).forEach(item => {
		templateString = templateString.replace(item[0].toUpperCase(), item[1])
	})
	return templateString;
}