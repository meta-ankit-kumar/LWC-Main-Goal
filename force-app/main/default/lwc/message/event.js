export class ClearMessageEvent extends CustomEvent {
	static type='clearmessage';
	constructor(value) {
		super(ClearMessageEvent.type, { detail: { value }});
	}
}