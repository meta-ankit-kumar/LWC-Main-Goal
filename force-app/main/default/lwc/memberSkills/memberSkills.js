import { getListUi } from 'lightning/uiListApi';
import { createRecord } from 'lightning/uiRecordApi';
import { api, LightningElement, track, wire } from 'lwc';
import { MESSAGE } from './constants';
import { MemberAddedEvent } from './event';
import { useTemplate } from './utils';

export default class MemberSkills extends LightningElement {
	@api teams;
	@track errors;
	@track message;
	// @wire(getListUi, { objectApiName: 'Team__c', listViewApiName: 'All' })
	// wiredTeams({ data, error }) {
	// 	if (data) {
	// 		this.teams = JSON.parse(JSON.stringify(data.records.records.map(record => ({
	// 			id: record.fields.Id.value,
	// 			name: record.fields.Name.value
	// 		}))))
	// 	} else if (error) {
	// 		this.errors = error.message;
	// 	}
	// }

	handleSubmit(event) {
		event.preventDefault();

		const formData = new FormData(event.target);
		const memberName = formData.get('member-name');
		const teamName = formData.get('member-team-name');
		const memberSkills = formData.get('member-skills');
		const fields = {
			Name: memberName,
			Skills__c: memberSkills,
			Team__c: teamName
		}
		createRecord({ apiName: 'TeamMember__c', fields })
			.then(result => {
				this.message = useTemplate(MESSAGE.MEMBER_CREATION.SUCCESSFULL, { MEMBER_NAME: result.fields.Name.value });
				this.dispatchEvent(new MemberAddedEvent({ bubbles: true }));
				const form = this.template.querySelector('form');
				form.reset();
			})
			.catch(error => {
				this.errors = error.message;
			});
	}

	handleClearMessage(event) {
		this.message = null;
	}

	get errorsToBeShown() {
		return useTemplate(MESSAGE.MEMBER_CREATION.ERROR, { ERRORS: this.errors });
	}
}