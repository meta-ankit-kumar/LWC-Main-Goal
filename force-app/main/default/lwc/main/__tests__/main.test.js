import { createElement } from 'lwc';
import Main from 'c/main';
import { getListUi } from 'lightning/uiListApi';
import { teamListData } from './data';
import { MemberAddedEvent } from '../../memberSkills/event';

describe('c-main', () => {
    afterEach(() => {
        // The jsdom instance is shared across test cases in a single file so reset the DOM
        while (document.body.firstChild) {
            document.body.removeChild(document.body.firstChild);
        }
    });

    it('should render the component successfully', () => {
        // GIVEN
        const element = createElement('c-main', {
            is: Main
        });

        // WHEN
        getListUi.emit(teamListData);
        document.body.appendChild(element);

        // THEN
        expect(element).toMatchSnapshot();
    });

    it('should handle the member change event', () => {
        // GIVEN
        const element = createElement('c-main', {
            is: Main
        });

        getListUi.emit(teamListData);
        document.body.appendChild(element);
        const memberSkillsComponent = element.shadowRoot.querySelector('c-member-skills');
        
        // WHEN
        memberSkillsComponent.dispatchEvent(new MemberAddedEvent());


        // THEN
        expect(element).toMatchSnapshot();
    });
});