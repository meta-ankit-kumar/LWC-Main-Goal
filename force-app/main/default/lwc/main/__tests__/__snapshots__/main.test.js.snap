// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c-main should handle the member change event 1`] = `
<c-main
  x-test_main-host=""
>
  #shadow-root(open)
    <div
      class="main-container"
      x-test_main=""
    >
      <c-member-skills
        x-test_main=""
        x-test_memberskills-host=""
      >
        #shadow-root(open)
          <div
            class="main-member-container"
            x-test_memberskills=""
          >
            <div
              class="member-skills-container"
              x-test_memberskills=""
            >
              <div
                class="labels-container"
                x-test_memberskills=""
              >
                <label
                  for="[shadow:guid]"
                  x-test_memberskills=""
                >
                  Name:
                </label>
                <label
                  for="[shadow:guid]"
                  x-test_memberskills=""
                >
                  Team:
                </label>
                <label
                  for="[shadow:guid]"
                  x-test_memberskills=""
                >
                  Skills:
                </label>
              </div>
              <div
                class="inputs-container"
                x-test_memberskills=""
              >
                <form
                  x-test_memberskills=""
                >
                  <input
                    id="[shadow:guid]"
                    minlength="4"
                    name="member-name"
                    required=""
                    type="text"
                    x-test_memberskills=""
                  />
                  <select
                    name="member-team-name"
                    required=""
                    x-test_memberskills=""
                  >
                    <option
                      value="a002t00000RfotWAAR"
                      x-test_memberskills=""
                    >
                      M14
                    </option>
                    <option
                      value="a002t00000RfotXAAR"
                      x-test_memberskills=""
                    >
                      M15
                    </option>
                    <option
                      value="a002t00000RfotbAAB"
                      x-test_memberskills=""
                    >
                      M16
                    </option>
                    <option
                      value="a002t00000RfotgAAB"
                      x-test_memberskills=""
                    >
                      M17
                    </option>
                  </select>
                  <input
                    id="[shadow:guid]"
                    name="member-skills"
                    required=""
                    type="text"
                    x-test_memberskills=""
                  />
                  <input
                    class="submit-button"
                    type="submit"
                    value="Submit"
                    x-test_memberskills=""
                  />
                </form>
              </div>
            </div>
          </div>
      </c-member-skills>
      <c-team-list
        x-test_main=""
        x-test_teamlist-host=""
      >
        #shadow-root(open)
          <div
            class="team-list-container"
            x-test_teamlist=""
          >
            <div
              class="team-selection-container"
              x-test_teamlist=""
            >
              <p
                class="team-selection-input-label"
                x-test_teamlist=""
              >
                Please select the Team: 
              </p>
              <select
                name="member-team-name"
                required=""
                x-test_teamlist=""
              >
                <option
                  value="a002t00000RfotWAAR"
                  x-test_teamlist=""
                >
                  M14
                </option>
                <option
                  value="a002t00000RfotXAAR"
                  x-test_teamlist=""
                >
                  M15
                </option>
                <option
                  value="a002t00000RfotbAAB"
                  x-test_teamlist=""
                >
                  M16
                </option>
                <option
                  value="a002t00000RfotgAAB"
                  x-test_teamlist=""
                >
                  M17
                </option>
              </select>
            </div>
            <div
              x-test_teamlist=""
            >
              <p
                x-test_teamlist=""
              >
                Please select a team to display the team members
              </p>
            </div>
            <div
              class="members-list"
              x-test_teamlist=""
            />
          </div>
      </c-team-list>
    </div>
</c-main>
`;
